tactics/ind_tables.ml

```diff
@@ -124,36 +124,66 @@ let local_lookup_scheme eff kind ind = match lookup_scheme kind ind with
 let local_check_scheme kind ind eff =
   Option.has_some (local_lookup_scheme eff kind ind)
 
-let define internal role id c poly univs =
+let define ?loc internal role id c poly univs =
   let id = compute_name internal id in
   let uctx = UState.minimize univs in
   let c = UState.nf_universes uctx c in
   let univs = UState.univ_entry ~poly uctx in
-  !declare_definition_scheme ~internal ~univs ~role ~name:id c
+  !declare_definition_scheme ~internal ~univs ~role ~name:id ?loc c
+
+  module Locmap : sig
+
+    type t
+
+    val default : Loc.t option -> t
+    val of_lidents : ?default:Loc.t -> MutInd.t -> Loc.t option list -> t
+    val lookup : locmap:t -> Names.inductive -> Loc.t option
+
+end = struct
+
+    type t = {
+      default : Loc.t option;
+      ind_to_loc : Loc.t Names.Indmap.t;
+    }
+    let lookup ~locmap:{ ind_to_loc; default } x =
+      Names.Indmap.find_opt x ind_to_loc |> fun loc ->
+      Option.append loc default
+
+    let default default = { default; ind_to_loc = Names.Indmap.empty }
+
+    let of_lidents ?default mind locs =
+      let default, ind_to_loc =
+        CList.fold_left_i (fun i (default,m) loc ->
+          Option.append default loc,
+          Option.cata (fun loc -> Names.Indmap.add (mind, i) loc m) m loc)
```

Comment:
I don't much like these Option combinators, especially `append` which is a nonsense name for something which forgets information (in the Some,Some case)
This may be easier to read:
```suggestion
          let m = match loc with
          | None -> m
          | Some loc -> Indmap.add (mind, i) loc m
          in
          let default = if Option.has_some default then default else loc in
          default, m
```